flink-core/src/main/java/org/apache/flink/core/memory/MemorySegmentFactory.java

```diff
@@ -94,7 +99,15 @@ public static MemorySegment allocateUnpooledOffHeapMemory(int size) {
 	 * @return A new memory segment, backed by unpooled off-heap memory.
 	 */
 	public static MemorySegment allocateUnpooledOffHeapMemory(int size, Object owner) {
-		ByteBuffer memory = ByteBuffer.allocateDirect(size);
+		ByteBuffer memory;
+		try {
+			memory = ByteBuffer.allocateDirect(size);
+		} catch (OutOfMemoryError e) {
+			if (ExceptionUtils.isDirectOutOfMemoryError(e)) {
+				LOG.error(ExceptionUtils.TM_DIRECT_OOM_ERROR_MESSAGE, e);
+			}
+			throw e;
+		}
```

Comment:
I was not sure whether this will always get into the Task thread error handling, sometimes it is called in Netty channel read handlers. I can do more research on it. Maybe, @zhijiangW has a quick good idea.

Reply:
Some inputs from my side: there are two usages for `allocateUnpooledOffHeapMemory` ATM.

- One is for initializing `NetworkBufferPool` in `TaskManagerServices`, and it also explicitly handles the `OutOfMemoryError` message inside. Maybe we also want to unify the message there.

- Another is for reading blocking partitions for batch jobs on upstream side, which is triggered by netty stack. This error would not be thrown outsides to be visible in task stack, and be only transported to the downstream side via network to cause the respective task failure. It also makes sense to handle this special error on upstream side, to make processor exit if possible. Otherwise the same error might happen again after failover recovery.

Reply:
For the second usage, it had a ticket (https://issues.apache.org/jira/browse/FLINK-15981) for improving this direct memory overhead,  not sure when will be done, just FYI.

Reply:
But if the receiver sees the exception, then the user will see it in the logs, right? This is what we want to achieve here. Making the sender handle the exception (e.g. shutting down or so) could then be a separate issue.

Reply:
I agree, improving the error message of exception should also help here to report more details to the receiver.
It would be nice to figure out a way to handle direct OOM errors from Netty threads in general because Netty also can allocate direct memory. Same for Netty threads in Akka RPC framework. These tasks can be follow-ups.

Reply:
> But if the receiver sees the exception, then the user will see it in the logs, right?
Yes.

Reply:
I suggest we keep this change to address this reported problem [FLINK-15981](https://issues.apache.org/jira/browse/FLINK-15981) and log it on sender side until we establish a general error reporting for the sender. The error message, got by receiver, is also improved now in a separate commit.